### PR TITLE
[MinBuild] 132KB minimal build binary size reduction via dummy __cxa_demangle

### PR DESCRIPTION
**Pull request by cloudhan**
**Description**:

bloaty shows following size contribution to the mininmal build binary
```
   2.2%  89.6Ki   2.1%  89.6Ki    /buildbot/src/android/ndk-release-r23/toolchain/llvm-project/libcxx/../../../toolchain/llvm-project/libcxxabi/src/cxa_demangle.cpp
```

Add the option `-Wl,--trace-symbol=__cxa_demangle`, shows `__cxa_demangle` is only referenced in
```
/home/guangyunhan/android_sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a: lazy definition of __cxa_demangle
/home/guangyunhan/android_sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a(cxa_default_handlers.o): reference to __cxa_demangle
/home/guangyunhan/android_sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a(cxa_demangle.o): definition of __cxa_demangle
```

It is possible to recompile the `libc++abi.a` with `LIBCXXABI_SILENT_TERMINATE` defined. But it might make the building too complicated at the moment.

To further reduce the binary size. We can provide a **dummy** implementation of `__cxa_demangle`

Do a minimal build without any operator included for `libonnxruntime.so`:

Before: 1109976
After:   977568

~132 KB reduction

**Motivation and Context**
- Why is this change required? What problem does it solve?
   Minimal build is currently reaching the limit. This change free us to add more useful code.

### CODE REVIEW

File: onnxruntime/core/so/cxa_demangle.cc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * However it contribute a large percentage of binary size in minial build. To avoid it, the LLVM user can compile
+ * libc++abi.a with LIBCXXABI_NON_DEMANGLING_TERMINATE defined, see https://reviews.llvm.org/D88189
+ * But this make the compilation process extremely complex. Here we provide a dummy __cxa_demangle. It always copy
+ * the mangled name to output buffer. The required allocation behavior in Itanium C++ ABI will result a failure status,
+ * a.k.a, not implemented.
+ */
+char* __cxa_demangle(const char* mangled_name, char* output_buffer, size_t* length, int* status) {
+  if (!status) return NULL;
+  if (!output_buffer) {
+    *status = -1;
```
Comment (skottmckay) [inline onnxruntime/core/so/cxa_demangle.cc:19]: nit: given we didn't do the memory allocation should this return -3 for invalid argument instead?

would be nice to have some names to the error codes for readability - either via an enum or some constants.

File: onnxruntime/core/so/cxa_demangle.cc [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    return NULL;
+  }
+  if (!mangled_name || !length || (*length == 0) || (*length > INT_MAX)) {
+    *status = -3;
+    return NULL;
+  }
+
+  int i = 0;
+  int limit = (int)(*length) - 1;
```
Comment (skottmckay) [inline onnxruntime/core/so/cxa_demangle.cc:28]: Avoid C-style casts. gsl::narrow<int>(*length) would be better (gsl::narrow_cast shouldn't be necessary here).

File: onnxruntime/core/so/cxa_demangle.cc [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  for (; i < limit; i++) {
+    output_buffer[i] = mangled_name[i];
+    if (mangled_name[i] == '\0') break;
+  }
```
Comment (skottmckay) [inline onnxruntime/core/so/cxa_demangle.cc:32]: strncpy_s

File: onnxruntime/core/so/cxa_demangle.cc [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#if USE_DUMMY_EXA_DEMANGLE
```
Comment (skottmckay) [inline onnxruntime/core/so/cxa_demangle.cc:1]: If we're only using this for android, onnxruntime/platform/android would be a better location rather than creating a new directory with an unclear name (not sure what 'so' is meant to stand for).

File: onnxruntime/core/so/cxa_demangle.cc (hunk 1 above)
Comment (cloudhan) [reply onnxruntime/core/so/cxa_demangle.cc:19]: In the API doc, the buffer can be null. From the caller's PoV, it is valid, so I think it is better to use `memory allocation failure` in this case.

Comment (skottmckay) [reply onnxruntime/core/so/cxa_demangle.cc]: If it's null isn't the implementation meant to allocate, which would potentially make returning an error about memory allocation valid? 

>If buf is a null pointer, __cxa_demangle allocates a new buffer with malloc. It stores the size of the buffer in *n, if n is not NULL.

File: onnxruntime/core/so/cxa_demangle.cc (hunk 4 above)
Comment (cloudhan) [reply onnxruntime/core/so/cxa_demangle.cc]: Moved

File: onnxruntime/core/platform/android/cxa_demangle.cc [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    return NULL;
+  }
+
+  // size is the buffer size, including the null char
+  size_t output_size = n ? *n : 0;
+  // It is caller's responsibility to ensure mangled_name is null-terminated as it is required in the API doc.
+  size_t input_size = strlen(mangled_name) + 1;
+  if (input_size > output_size) {
+    free(buf);
+    buf = NULL;
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/platform/android/cxa_demangle.cc:36]: See the last line of the ABI doc, it says:

If buf is not big enough to store the resulting demangled name, __cxa_demangle must _either_ **a) call free to deallocate buf and then allocate a new buffer with malloc**, or b) call realloc to increase the size of the buffer. In either case, the new buffer size will be stored in *n.

wrt the cxa_default_handlers.cpp, I think it is the implementor is sure about the symbol length is always less than 1024?

[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/platform/android/cxa_demangle.cc:36]: Maybe they are sure it's always less than 1024. However, given that usage of passing in a pointer to a stack allocated buffer I don't think it's safe for the implementation to assume it can call `free` (despite what the documentation says) as that will crash at runtime. The documentation isn't going to stop it crashing :P.

File: onnxruntime/core/platform/android/cxa_demangle.cc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#include <limits.h>
+
+#define DEMANGLE_INVALID_ARGUMENTS -3
+#define DEMANGLE_INVALID_MANGLED_NAME -2
+#define DEMANGLE_MEMORY_ALLOCATION_FAILURE -1
+#define DEMANGLE_SUCCESS 0
+
+// Reduced from https://github.com/llvm/llvm-project/blob/09c2b7c35a/libcxxabi/src/demangle/Utility.h
+inline bool initialize_output_buffer(char*& buf, size_t* n, size_t& buf_size, size_t init_size) {
+  // only handle buf == null case, is buf is provided, in original impl, it realloc to expand buf space.
```
Comment (skottmckay) [inline onnxruntime/core/platform/android/cxa_demangle.cc:14]: static

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Comment by cloudhan** [comment]
I am adding some tests, but gtest ASSERT_DEATH is causing flaky behaviour :(

**Comment by skottmckay** [comment]
> I am adding some tests, but gtest ASSERT_DEATH is causing flaky behaviour :(

Can you test by calling `__cxa_demangle` directly? Don't necessarily need to go via a path that exits/throws to test the implementation.

File: onnxruntime/core/platform/android/cxa_demangle.cc (hunk 6 above)
Comment (cloudhan) [reply onnxruntime/core/platform/android/cxa_demangle.cc:17]: pointer is only guarantee to be non-null iff buf is non-null

File: tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml [hunk 7]
```diff
@@ -261,6 +261,7 @@ jobs:
         --android_abi=x86_64 \
         --android_api=29 \
         --use_nnapi \
+        --build_shared_lib \
         --test
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:265]: Out of interest why was this change required?

Comment (cloudhan) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:265]: This change is only affect a shared library, so the test is attached to onnxruntime_shared_libs_test, so this is enabled.

Comment (cloudhan) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml]: This line is reverted after a rebase and squash.

File: tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml [hunk 8]
```diff
@@ -358,6 +358,7 @@ jobs:
         --skip_submodule_sync \
         --parallel \
         --use_nnapi \
+        --build_shared_lib \
         --cmake_generator=Ninja \
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:362]: Is this change needed? Wondering if there's any implication on the code coverage results if we start building a shared lib instead.

Comment (cloudhan) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:362]: Ouch, that might be a case, indeed. So I think I should move the test to `onnxruntime_test_all` target instead? So that we don't need the `--build_shared_lib` option

Comment (skottmckay) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:362]: I don't know for sure. As we're only plugging in the dummy demangle if the shared library is being built, moving the test has other implications. I'd try a temporary change in the CI config to comment out the `condition` so it runs code coverage for all builds to see if there is any impact on it.

Comment (cloudhan) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:362]: Elaborate?

> comment out the `condition`

Comment (skottmckay) [reply tools/ci_build/github/azure-pipelines/android-x86_64-crosscompile-ci-pipeline.yml:362]:
```yaml
# The below jobs only run on master build
- job: NNAPI_EP_MASTER
  pool:
    vmImage: 'macOS-11'
  timeoutInMinutes: 120
  workspace:
    clean: all
  condition: in(variables['Build.Reason'], 'IndividualCI', 'BatchedCI')
```

Last line there determines if the job runs. Comment that out temporarily so it should run for all builds.

**Review by skottmckay — APPROVED** [review]
:shipit:
